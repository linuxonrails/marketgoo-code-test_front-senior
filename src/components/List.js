import React from "react";
import axios from "axios";
import { Button } from "@marketgoo/ola";

class List extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            data: null,
            name: null,
            team: null,
            score: null,
        };
        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleDelete = this.handleDelete.bind(this);
        // Create an AbortController to cancel API requests:
        this.controller = new AbortController();
    }

    componentDidMount() {
        axios
            .get("/players", { signal: this.controller.signal })
            .then((response) => {
                this.setState({ data: response.data });
            })
            .catch((err) => {
                if (err.name == "CanceledError") {
                    // Canceled API request: do nothing
                } else {
                    throw err;
                }
            });
    }

    componentWillUnmount() {
        this.controller.abort();
    }

    handleChange(event) {
        this.setState({ [event.target.name]: event.target.value });
    }

    handleSubmit(event) {
        event.preventDefault();
        const { name, team, score } = this.state;
        axios.post("/players", { name, team, score }).then(() => {
            axios.get("/players").then((response) => {
                this.setState({ data: response.data });
            });
        });
    }

    handleDelete(id) {
        axios.delete(`/players/${id}`).then(() => {
            axios.get("/players").then((response) => {
                this.setState({ data: response.data });
            });
        });
    }

    render() {
        const players = this.state.data;

        return (
            <>
                <table>
                    <thead>
                        <tr>
                            <td>Player</td>
                            <td>Team</td>
                            <td>Score</td>
                            <td>Actions</td>
                        </tr>
                    </thead>
                    <tbody data-testid="players">
                        {players &&
                            players.data.map((x) => (
                                <tr key={x.id} data-testid="player">
                                    <td>{x.name}</td>
                                    <td>{x.team}</td>
                                    <td>{x.score}</td>
                                    <td>
                                        <button
                                            onClick={() =>
                                                this.handleDelete(x.id)
                                            }
                                        >
                                            Remove
                                        </button>
                                    </td>
                                </tr>
                            ))}
                    </tbody>
                </table>
                <br />
                <form onSubmit={this.handleSubmit}>
                    <h4>Add new players</h4>
                    <label htmlFor="name" style={{ display: "none" }}>
                        Player name
                    </label>
                    <input
                        id="name"
                        name="name"
                        placeholder="player name"
                        onChange={this.handleChange}
                    ></input>
                    <label htmlFor="team" style={{ display: "none" }}>
                        Team name
                    </label>
                    <input
                        id="team"
                        name="team"
                        placeholder="team name"
                        onChange={this.handleChange}
                    ></input>
                    <label htmlFor="score" style={{ display: "none" }}>
                        Team score
                    </label>
                    <input
                        id="score"
                        name="score"
                        placeholder="team score"
                        onChange={this.handleChange}
                    ></input>
                    <Button variant="primary">Add</Button>
                </form>
            </>
        );
    }
}

export default List;
