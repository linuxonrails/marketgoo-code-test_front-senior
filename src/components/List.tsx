import React from "react";
import axios from "axios";
import { PlayerTable } from "./PlayerTable";
import { FormAddNewPlayers } from "./FormAddNewPlayers";

type Player = {
    id: number;
    name: string;
    score: number;
    team: string;
    updateAt: string;
};

type PlayerAPIResponse = {
    status: "ok";
    data: Player[];
};

type NewPlayer = {
    name: string;
    score: string;
    team: string;
};

export const List: React.FC<{}> = ({}) => {
    const [players, setPlayersTo] = React.useState<Player[]>(null);
    const [new_player, setNewPlayerTo] = React.useState<NewPlayer>(null);

    // Create an AbortController to cancel API requests:
    const controller = new AbortController();

    React.useEffect(() => {
        console.log("componentDidMount");
        axios
            .get<PlayerAPIResponse>("/players", { signal: controller.signal })
            .then((response) => {
                setPlayersTo(response.data.data);
            })
            .catch((err) => {
                if (err.name == "CanceledError") {
                    // Canceled API request: do nothing
                } else {
                    throw err;
                }
            });
    }, []);

    const componentWillUnmount = () => {
        controller.abort();
    };

    const handleChange = (event) => {
        setNewPlayerTo({
            ...new_player,
            [event.target.name]: event.target.value,
        });
    };

    const handleSubmit = (event) => {
        event.preventDefault();
        const { name, team, score } = new_player;
        axios.post("/players", { name, team, score }).then(() => {
            axios.get("/players").then((response) => {
                setPlayersTo(response.data.data);
            });
        });
    };

    const handleDelete = (id) => {
        axios.delete(`/players/${id}`).then(() => {
            axios.get("/players").then((response) => {
                setPlayersTo(response.data.data);
            });
        });
    };

    console.log({ players });

    return (
        <>
            <PlayerTable players={players} handleDelete={handleDelete} />
            <br />
            <FormAddNewPlayers
                handleChange={handleChange}
                handleSubmit={handleSubmit}
            />
        </>
    );
};

export default List;
