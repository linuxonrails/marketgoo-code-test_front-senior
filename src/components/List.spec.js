import React from "react";
import { render, screen } from "@testing-library/react";
import List from "./List";

describe("List", () => {
    it("should render table with a player list", () => {
        // Arranage
        render(<List />);
        // Act
        screen.getByText("Player");
        screen.getByText("Score");
        screen.getByText("Actions");
    });

    it("should render form to add new players", () => {
        // Arranage
        render(<List />);
        // Act
        screen.getByText("Add new players");
        screen.getByRole("button", { name: "Add" });
    });
});
