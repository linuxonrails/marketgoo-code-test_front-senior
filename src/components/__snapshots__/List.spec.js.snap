// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should match the same snapshot 1`] = `
<DocumentFragment>
  <table>
    <thead>
      <tr>
        <td>
          Player
        </td>
        <td>
          Team
        </td>
        <td>
          Score
        </td>
        <td>
          Actions
        </td>
      </tr>
    </thead>
    <tbody
      data-testid="players"
    />
  </table>
  <br />
  <form>
    <h4>
      Add new players
    </h4>
    <label
      for="name"
      style="display: none;"
    >
      Player name
    </label>
    <input
      id="name"
      name="name"
      placeholder="player name"
    />
    <label
      for="team"
      style="display: none;"
    >
      Team name
    </label>
    <input
      id="team"
      name="team"
      placeholder="team name"
    />
    <label
      for="score"
      style="display: none;"
    >
      Team score
    </label>
    <input
      id="score"
      name="score"
      placeholder="team score"
    />
    <button
      class="ola_button is-primary"
    >
      <span
        class="ola_button-text"
      >
        Add
      </span>
    </button>
  </form>
</DocumentFragment>
`;
